### ISSUE TITLE
[iOS] CollectionView item cells do not resize when the content size is changed

### ISSUE THREAD

**Issue by Alex-Dobrynin** [dotnet/maui#21141]
### Description

I have collectionview and itemtemplate that can change it's size. Unfortunately, the collectionview item cell do not change it's size.
In Xamarin.Forms this is not an issue

[CollectionViewTest.zip](https://github.com/dotnet/maui/files/14564470/CollectionViewTest.zip)

### Steps to Reproduce

1. Build and Run attached project
2. click on items
3. see the items are not resized

### Version with bug

8.0.7 SR2

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Did you find any workaround?

nope

### Relevant log output

```shell
no logs
```

**Comment by mattleibow** [comment]
I see you have a ContentView in the items, can you try removing that and seeing if the layouts update?

**Comment by Alex-Dobrynin** [comment]
@mattleibow 
that didnt help, also i've already tried different containers as grid or stack, even just expander without any container, does not work

I even debugged maui repo to find out why so. and found that inside TemplatedCell the MeasureInvalidated event is not being raised after size of content changed.

By the way, you said that I have ContentView as root of template. It should also change it's size, when content size changed, in Xamarin.Forms this is not an issue, everything works as expected

**Comment by joxi** [comment]
Probably the same problem as reported in https://github.com/dotnet/maui/issues/14363
I agree, this is very important and a real pain to workaround.

**Comment by Larhei** [comment]
We are also facing this. For example when changing the visibility of some part in the template after cell was rendered. In case the contentsize would increase on ios i see a lot of rendering issues.

**Comment by DiogoDeAlmeida** [comment]
Hi guys !

We are facing the same issue (With an Expander into a CollectionView) and after many attempts to find out a workaround i have tried to use BindableLayout instead of CollectionView/ListView or even DXCollectionView and its seems to work as expected. I don't know if it could help you to find a solution but it look like this :

```CSharp

 <VerticalStackLayout BindableLayout.ItemsSource="{Binding MyList}">
     <BindableLayout.ItemTemplate>
         <DataTemplate x:DataType="model:MyModel">
                   <toolkit:Expander>
                             <toolkit:Expander.Header>
                                  <Label Text="{Binding Title}" />
                             </toolkit:Expander.Header>
                             <Label Text="{Binding Content}" />
                    </toolkit:Expander>
        </DataTemplate>
    </BindableLayout.ItemTemplate>
</VerticalStackLayout>
```

**Comment by Alex-Dobrynin** [comment]
@DiogoDeAlmeida using stack with bindable layout is absolutely wrong, when you have long list, this is because you need virtualization if you don't want to get out of memory exception

**Comment by DiogoDeAlmeida** [comment]
@Alex-Dobrynin i agree with you about the usage of virtualization to optimize the navigation and avoid memory exception, if you have a lot of elems in your list don't use BindableLayout ! My purpose here is to show the possibility to achieve this issue (having a list with items wich can change their size)

In my use case i can't change the design system for iOS, so i tried to figure out how i could achieve it without change everything, i have a short list (<10 elements) and it works perfectly so it's not wrong when you try to achieve something workable

We will fix it as soon as MAUI will be patched for raising the SizeChanged event properly, until that (and it's seems to be an issue for a long now) we try to find workaround

**Comment by claytonone** [comment]
We've run into this issue as well. It effectively makes CollectionView unusable on iOS if you have items that resize. Any known work arounds?

**Comment by aherrick** [comment]
forcing a call to PlatformSizeChanged(); on iOS has "fixed" this for me

**Comment by Alex-Dobrynin** [comment]
@aherrick this is ok, when you know when the resize is happend, but what if you dont know when, or some property of DataTemplate's bindingcontext changed. This should not be done manually

**Comment by aherrick** [comment]
@Alex-Dobrynin agreed! just pointing out i was able to get a work around in place

**Comment by claytonone** [comment]
> forcing a call to PlatformSizeChanged(); on iOS has "fixed" this for me

Out of interest how did you do that?

**Comment by aherrick** [comment]
@claytonone its not pretty, but just literally forcing a call to PlatformSizeChanged(); in my data template item frame for the collection view. I'm responding to a button click to show more data in the cell, then once that data is show i force PlatformSizeChanged();

**Comment by mariomoramontero** [comment]
Hi
I'm using an Expander in a CollectionView but sadly the "PlatformSizeChanged();" don't work for me. After many tries and lots of wasted time I could achieve it by calculating manually the height of the child elements and assigning HeightRequest on ExpanderChanged event

**Comment by tentom** [comment]
Don't know if it will help with fixing this issue but my content is resizing if the element is scrolled out of the screen and then back in again.

I have found no workaround yet.

**Comment by FlavioGoncalves-Cayas** [comment]
Will we ever see this resolved? Is anyone actively working on this? It's not like this is some edge case that affects only 0.1% of apps, so that it can be disregarded as much as it is...

**Comment by billylo1** [comment]
Android doesn't have this problem, BTW.

**Comment by arosanti** [comment]
I have come up with a not great but effective workaround for this issue
`private void TapGestureRecognizer_Tapped(object sender, EventArgs e)
{
    if (sender is Element element && element.BindingContext is EquipmentWrapperModel equipmentModel)
    {
        // Toggle the IsExpanded property
        _viewModel.ExpandEquipmentCommand.Execute(equipmentModel);

        if (DeviceInfo.Platform == DevicePlatform.iOS)
        {
            _viewModel.InstalledEquipmentList = new List<EquipmentWrapperModel>(_viewModel.InstalledEquipmentList);
            equipmentList.ScrollTo(equipmentModel);
        }                
    }            
}`

**Comment by Baraiboapex** [comment]
> Hi I'm using an Expander in a CollectionView but sadly the "PlatformSizeChanged();" don't work for me. After many tries and lots of wasted time I could achieve it by calculating manually the height of the child elements and assigning HeightRequest on ExpanderChanged event

Do you mind showing the code for this? I tried something like this but couldn't figure it out.

**Comment by mariomoramontero** [comment]
> @mariomoramontero
> 
> > Hi I'm using an Expander in a CollectionView but sadly the "PlatformSizeChanged();" don't work for me. After many tries and lots of wasted time I could achieve it by calculating manually the height of the child elements and assigning HeightRequest on ExpanderChanged event
> 
> Do you mind showing the code for this? I tried something like this but couldn't figure out how to get the height of each child element within the expander. Did you have fixed height for your items or were they all automatically sized? Getting the sizes of an element that is automatically sized and within the children of a collection view is what I need for this to work.

Sorry for the late, didn't see that... sure, I can paste my code

```
private void Expander_ExpandedChanged( Object sender, ExpandedChangedEventArgs e )
{
   //TODO: this is an ugly workaround only for iOS
   //https://github.com/dotnet/maui/issues/21271
   //https://github.com/dotnet/maui/issues/21141
   //please review after some time
   if( DeviceInfo.Current.Platform == DevicePlatform.iOS )
   {
      var expander = sender as Expander;     

      if( expander?.Height > -1 )
      {
         double height = ( expander.Content as Grid )?.Children.Sum( p => p.Height ) ?? 0;

         if( e.IsExpanded )
         {
            ( expander.Parent as View )!.HeightRequest = height;
         }
         else
         {
            ( expander.Parent as View )!.HeightRequest = 40;
         }
      }
   }
}
```

**Comment by billylo1** [comment]
> Here's another (bad) workaround:
> 
> ```
> if (DeviceInfo.Current.Platform == DevicePlatform.iOS)
> {
>      collectionView.ItemSizingStrategy = ItemSizingStrategy.MeasureFirstItem;
>      collectionView.ItemSizingStrategy = ItemSizingStrategy.MeasureAllItems;
> }
> ```
> 
> It's very slow, but does the job.

thank you for sharing.  This worked for me too and I have adopted it for now...

**Comment by maccam04** [comment]
We've also encountered this problem. We have a show/hide control in the template, and when the toggle is on, a control will show up, but it will overlap with the next item.

I have tested it with MAUI Service Release 7 and .NET 8 nightly feed (MAUI version 8.0.80-ci.net8.24373.1), but the result is the same. 

Here's my MAUI repro project. 
[CollectionViewResizeBug.zip](https://github.com/user-attachments/files/16368110/CollectionViewResizeBug.zip)

**Comment by michele-guion** [comment]
+1

**Comment by LeoJHarris** [comment]
Any update?? So far it certainly seems Xamarin is much more stable then MAUI right now. Running into a few UI glitches since beginning our app migration.

**Comment by FlavioGoncalves-Cayas** [comment]
> Any update?? So far it certainly seems Xamarin is much more stable then MAUI right now. Running into a few UI glitches since beginning our app migration. Will progress some other issues and hope to loop back and find this issue resolved by the MAUI team 🤞

This has never properly worked ever in .NET MAUI, neither when MAUI went GA more than 2 years ago nor with .NET 9 coming up. And the worst part of it is that the MAUI team does not seem to be interested in fixing this.

**Comment by Baraiboapex** [comment]
I hate to say it guys, but they are using the excuse that since this is an "open source project", we need to be the ones to fix this problem. It sucks, I know but it is what it is.

**Comment by LeoJHarris** [comment]
After doing some digging on a replacement for `CollectionView` I've started using @Alex-Dobrynin very own `MPowerKit.VirtualizeListView` found here https://github.com/MPowerKit/VirtualizeListView and its working like a charm, if anyone else is looking for solution and waiting on MAUI team to resolve this issue I recommend this for now.

**Comment by arosanti** [comment]
@PureWeen I just tried using the latest nightly build and my issue was fixed.

**Comment by Alex-Dobrynin** [comment]
@PureWeen yes, but it does proper size calculation only from second expand of item
watch attached video

https://github.com/user-attachments/assets/25802ad4-9656-4bf5-b476-2d87e33a903d

**Comment by albyrock87** [comment]
@Alex-Dobrynin are you sure this is not a problem with Toolkit Expander?
What if you use `IsVisible` or `MaxHeightRequest` instead?

**Comment by Alex-Dobrynin** [comment]
@albyrock87 I'm sure, because in not using it

**Comment by albyrock87** [comment]
@Alex-Dobrynin then may you share a code to reproduce?
Your first post here is using the Expander from mct.

**Comment by Alex-Dobrynin** [comment]
@albyrock87 simply just using custom expander based on templatedview, where controltemplate is just grid with 2 rows, each row contains contentview as header and content. For 2 row just changing the visibility depending on expander's state. Thats all

**Comment by albyrock87** [comment]
@Alex-Dobrynin I tried to do exactly what you suggested and I cannot reproduce (source code attached in the ZIP file below).

![Simulator Screen Recording - iPhone 15 Pro Max - 2024-09-07 at 11 05 32](https://github.com/user-attachments/assets/737f124e-e339-4d92-bbbe-679c2894545f)

[Issue21141.test.zip](https://github.com/user-attachments/files/16917182/Issue21141.test.zip)

**Comment by Alex-Dobrynin** [comment]
@albyrock87 here you can find very simplified layout i have in my app, sorry but i cannot provide you more detailed layout.

[Issue21141.zip](https://github.com/user-attachments/files/16917375/Issue21141.zip)

**Comment by albyrock87** [comment]
@Alex-Dobrynin wait a minute, you have a vertical collection view within another vertical collection view, and I think that is not supported.
May you replace the inner one with a BindableLayout and let me know if that solves the issue?

**Comment by Alex-Dobrynin** [comment]
@albyrock87 I sucessfully use this scenario in MPowerKit.VirtualizeListView and everything expands and works as expected. I need such thing as wraplayout. Therefore i used collectionview with gridlayout. If there is such control to wrap items like CV with grid layout does i will replace it

**Comment by albyrock87** [comment]
@Alex-Dobrynin this is a workaround for you which works on the latest nightly.

```
<CollectionView Margin="10,0"
                                        PropertyChanged="CollectionView_PropertyChanged"
                                        ItemsSource="{Binding SubItems}">
```


```
    private void CollectionView_PropertyChanged(object? sender, PropertyChangedEventArgs e)
    {
        if (e.PropertyName == nameof(CollectionView.ItemsSource) && sender is CollectionView collectionView)
        {
            // `CollectionView` is inside a `VerticalStackLayout` which means it lies in an infinite vertical space
            // Measure of `CollectionView` then depends on its content, but right now we just changed `ItemSource`
            Dispatcher.Dispatch(() =>
            {
                // During this main-loop the `CollectionView` will be rendered, so on the next loop we can read the actual measure
                Dispatcher.Dispatch(() =>
                {
                    // Now invalidate the measure of the `CollectionView` so that its container can re-evaluate the layout
                    collectionView.InvalidateMeasureNonVirtual(InvalidationTrigger.Undefined);
                });
            });
        }
    }
```

**Comment by albyrock87** [comment]
@Alex-Dobrynin I've opened a PR which should fix this remaining issue: once I have a build available I'll give it to you so you can confirm.

**Comment by Alex-Dobrynin** [comment]
You are closing issue even not asking the reporter if it was fixed? Highest level of competence, really.

**Comment by albyrock87** [comment]
@Alex-Dobrynin AFAIK that's an automatic process which happens once the PR has been merged.

You can try the latest nightly and confirm if everything is working now.

If it's still not working you can open a new issue specifically to target the nested CollectionView issue.

Thanks!

**Comment by LeoJHarris** [comment]
@Alex-Dobrynin I concur, that's the problem honestly with so many issues closed that they are actually NOT resolved, just dig through a few a few and you'll see "Still seeing this issue" or "Not resolved in MAUI version X.XXXX.X". 

> If it's still not working you can open a new issue specifically to target the nested CollectionView issue.

@albyrock87 Nobody wants to go through the same process of submitting a new issue when we already have an existing one, not to mention a sample repo was provided here, why duplicate the same issue elsewhere??? That just slows the whole process down.

**Comment by albyrock87** [comment]
@LeoJHarris my suggestion to open a new issue is because this one started complaining about something which is different from what we've been discussing in the last few comments.

- initial issue: CV cells not reacting to content size changes
- recent comments: CV cells not sizing properly with nested CV

The initial issue has been fixed *and confirmed* a few days ago, so this issue should have already been closed and instead I made an additional effort trying to also cover this use case.

**Comment by Alex-Dobrynin** [comment]
@albyrock87 this is the same issue, because nested CV is the cell's content also. For me, and for others there is no any difference what was put into the cell grid, contentview, label or another collectionview, all of those controls are `typeof(View)`

**Comment by albyrock87** [comment]
The difference is: `CollectionView` is a scrollable area.
Nesting scrollable areas with the same orientation is an anti-pattern and in general a very bad UX design.

In MAUI docs on `ScrollView` there is also a warning saying that nesting scrollable areas is not recommended/supported. This is mostly the reason why I'm saying that's is a different issue.

As you mentioned previously, you're using CV just because it has this convenient grid layout, but you're not using it as a scrollable component.
I think you should use a simple `Grid` with `BindableLayout` if you manage to put a row/column index in the view model.

Anyway, as a community contributor, I've already spent *a lot of time* to fix this issue for you all :) so I'm not gonna spend other time in the tentative to clarify the situation.

Please just let me know if my last PR fixes also the CV nesting use case, thanks!

**Comment by FlavioGoncalves-Cayas** [comment]
I can understand @Alex-Dobrynin frustration, but nonetheless a huge thanks to @albyrock87 for fixing this issue. You're awesome.
